### PR TITLE
fix: save inline edit block on outside click

### PR DESCRIPTION
**Pull request by storyworks**
# Description

Open for anyone to review - this is a cooldown 5 task I picked up while waiting for other work to be reviewed

[Save text content changes on when clicking outside the block](https://3.basecamp.com/3105655/buckets/28141725/todos/5055615357)

# How should this PR be QA Tested?

- [ ] Manually test below on Typography, Button, RadioOption & SignUp blocks on mobile and desktop
> - [ ] Edit text, click **outside the card**, text should save
> - [ ] Edit text, click **outside the element** (but still inside the card), text should save
> - [ ] Edit text on card **without a background**, click Done button, text should save
> - [ ] Edit text on card **with a background image**, click Done button, text should save
> - [ ] Edit text on card **with a background video**, click Done button, text should save

# Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged into main

### CODE REVIEW

File: apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/ButtonEdit/ButtonEdit.spec.tsx [hunk 1]
```diff
@@ -84,4 +84,57 @@ describe('ButtonEdit', () => {
     fireEvent.blur(input)
     await waitFor(() => expect(result).toHaveBeenCalled())
   })
+  it('saves the button label on blur', async () => {
```
Comment (storyworks) [inline apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/ButtonEdit/ButtonEdit.spec.tsx:87]: Oops will fix name.

File: apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/ButtonEdit/ButtonEdit.spec.tsx [hunk 2]
```diff
@@ -84,4 +84,57 @@ describe('ButtonEdit', () => {
...
+            </iframe>
+          </EditorProvider>
+        </JourneyProvider>
+      </MockedProvider>
+    )
+
+    const input = getByRole('textbox')
+    fireEvent.click(input)
+    fireEvent.change(input, { target: { value: '    updated label    ' } })
+    fireEvent.click(getByRole('heading', { level: 1 }))
```
Comment (storyworks) [inline apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/ButtonEdit/ButtonEdit.spec.tsx:137]: All these tests are essentially the same as the `save the block on blur` tests except for this line & I wrap the Editable block in an iframe to simulate what happens with our Canvas component

File: apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/useOnClickOutside/useOnClickOutside.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// div, heading, iframe, paragraph
```
Comment (storyworks) [inline apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/useOnClickOutside/useOnClickOutside.spec.tsx:1]: Oops will remove this too

**Review by tataihono — APPROVED** [review]

**Comment by storyworks** [comment]
Awaiting manual testing

File: apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/useOnClickOutside/useOnClickOutside.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+export function useOnClickOutside<T extends HTMLElement = HTMLElement>(
+  callback: () => void | Promise<void>
+): RefObject<T> {
+  const elementRef = useRef<T>(null)
+  useEffect(() => {
+    const handleClick = (event: MouseEvent): void => {
+      if (
+        elementRef.current != null &&
+        !elementRef.current.contains(event.target as HTMLElement) &&
+        elementRef.current.classList.contains('Mui-focused')
```
Comment (csiyang) [inline apps/journeys-admin/src/components/Editor/Canvas/InlineEditWrapper/useOnClickOutside/useOnClickOutside.tsx:12]: @mikeallisonJS the done button bug was a bit of a weird one. 

When there is a background image/video, useOnClickOutside, gets called twice, once with the correct changes and then again with the old text. This caused the text to save as the updated text and then immediately get saved as the old text.

The solution I came up with was to add an additional filter to the ref check. Would love to get your thoughts on this solution before getting QA on it.

**Pull request merged**